ucw/doc/config.txt
Configuration files
===================

This document describes run-time configuration of libucw-based
programs using config files. For compile-time configuration,
see <<configure:>>.

[[terminology]]
Terminology
-----------

Configuration items of all modules are organized into sections.
The sections form a tree structure with top-level sections corresponding
to program modules.

Each configuration item belongs to one of the following classes:

  1. single value or a fixed-length array of values
  2. variable-length array of values
  3. subsection with several nested attributes
  4. list of nodes, each being an instance of a subsection
  5. bitmap of small integers (0..31) or fixed list of strings
  6. exceptions (items with irregular syntax; however, they always
     appear as a sequence of strings, only the semantics differ)

Both fixed- and variable-length arrays consist of items of the same
type.  The basic types supported by the configuration mechanism are:

  1. 32-bit integer
  2. 64-bit integer
  3. floating point number
  4. IP address
  5. string
  6. choice (one of a fixed list of strings)

Program modules can define their own special types (such as network
masks or attribute names) and decide how are they parsed.

[[format]]
Format of configuration files
-----------------------------

Configuration files are text files that usually set one attribute per
line, though it is possible to split one assignment into multiple lines
and/or assign several attributes in one line.  The basic format of an
assignment command is

  name value1 value2 ... valueN

or

  name=value1 value2 ... valueN

The end of line means also end of a command unless it is preceded by a
backslash.  On the other hand, a semicolon terminates the command and
another command can start after the semicolon.  A hash starts a comment
that lasts until the end of the line.  A value can be enclosed in
apostrophes or quotation marks and then it can contain spaces and/or
control characters, otherwise the first space or control character
denotes the end of the value.  Values enclosed in quotation marks are
interpreted as C-strings.  For example, the following are valid
assignment commands:

  Database "main db\x2b"; Directory='index/'; Weights 100 20 30 \
    40 50 80				# a comment that is ignored

Numerical values can be succeeded by a unit.  The following units are
supported:

[[units]]

	d=86400		k=1000		K=1024
	h=3600		m=1000000	M=1048576
	%=0.01		g=1000000000	G=1073741824

Attributes of a section or a list node can be set in two ways.  First,
you can write the name of the section or list, open a bracket, and then
set the attributes inside the section.  For example,

  Section1 {
    Attr1	value1
    Attr2	value2
    ListNode {		#creates a list and adds its first node
      Attr3	value3
      Attr4	value4
    }
    ListNode { Attr3=value5; Attr4=value6 }
			#appends a new node; this is still the same syntax
  }

The second possibility is using a shorter syntax when all attributes of a
section are set on one line in a fixed order.  The above example could
be as well written as

  Section1 {
    Attr1	value1
    Attr2	value2
    ListNode	value3 value4
    ListNode	value5 value6
  }

Of course, you cannot use the latter syntax when the attributes allow
variable numbers of parameters.  The parser of the configuration files
checks this possibility.

If you want to set a single attribute in some section, you can also
refer to the attribute as Section.Attribute.

Lists support several operations besides adding a new node.  You just
have to write a colon immediately after the attribute name, followed by
the name of the operation.  The following operations are supported:

[[operations]]

  List:clear				# removes all nodes
  List:append { attr1=value1; ... }	# adds a new node at the end
  List:prepend { attr1=value1; ... }	# adds a new node at the beginning
  List:remove { attr1=search1 }		# find a node and delete it
  List:edit { attr1=search1 } { attr1=value1; ... }
					# find a node and edit it
  List:after { attr1=search1 } { ... }	# insert a node after a found node
  List:before { attr1=search1 } { ... }	# insert a node before a found node
  List:copy { attr1=search1 } { ... }	# duplicate a node and edit the copy
  List:reset { attr=value1; ... }	# equivalent to :clear and :append

You can specify several attributes in the search condition and the nodes
are tested for equality in all these attributes.  In the editing
commands, you can either open a second block with overridden attributes,
or specify the new values using the shorter one-line syntax.

The commands :clear, :append, and :prepend are also supported by var-length
arrays.  The command :clear can also be used on string values.  The following
operations can be used on bitmaps: :set (which is equal to :append and :prepend),
:remove, :clear, and :all (set all bits).

[[include]]
Including other files
---------------------

To include another file, use the command

  Include another/file

or if the file needs not to exist

  OptionalInclude another/file

(Beware that this command has to be the last one on the line.)

[[command_line]]
Command-line parameters
-----------------------

The default configuration file (cf_def_file possibly overriden
by environment variable cf_env_file) is read before the program is started.
You can use a -C option to override the name of the configuration file.
If you use this parameter several times, then all those files are loaded
consecutively. A parameter -S can be used to execute a configuration
command directly (after loading the default or specified configuration
file).  Example:

  bin/program -Ccf/my-config -S'module.trace=2;module.logfile:clear' ...

If the program is compiled with debugging information, then one more
parameter `--dumpconfig` is supported.  It prints all parsed configuration
items and exits.

All these switches must be used before any other parameters of the
program.

[[preprocess]]
Preprocessing
-------------

During compilation, all configuration files are pre-processed by a simple
C-like preprocessor, which supports `#ifdef`, `#ifndef`, `#if`,
`#elsif`, `#else` and `#endif` directives referring to compile-time
configuration variables (the ones detected by `configure` script, you
can see list of them in `obj/autoconf.h`). `#if` and `#elsif` can contain
any Perl expression where each `CONFIG_xyz` configuration variable is
substituted to 0 or 1 depending on its value.

The preprocessor also substitutes `@VARIABLE@` by the value of the variable,
which must be defined.

[[caveats]]
Caveats
-------

Trying to access an unknown attribute causes an error, but unrecognized
top-level sections are ignored.  The reason is that a common config file
is used for a lot of programs which recognize only their own sections.

Names of sections, attributes and choices are case-insensitive. Units are
case-sensitive.


ucw/doc/configure.txt
How to Configure Sherlock libraries
===================================

What can be configured
----------------------
There are two different levels of configuring/customizing program
based on sherlock libraries:

  - runtime configuration in configuration files (see <<config:>>)

  - compile-time configuration of the libraries: config switches set
    before compiling, selecting optional features.

Where to build
--------------
If you run configure in the source directory, it prepares for compilation inside
the source tree. In this case, an `obj` subdirectory is created to hold all generated
files (object files, binaries, generated source files etc.) and all final files
are linked to the `run` subdirectory. No other parts of the source tree are written into.

Alternatively, you can compile in a separate object tree (which is useful when you
want to build several different configurations from a single source tree). In order
to do that, switch to the destination directory and issue `<source-dir>/configure ...`.
This way, configure will create the `obj` and `run` directories locally and set up
a Makefile which refers to the original source tree.

How to configure
----------------
To set up compilation, possibly overriding default compile-time
options, just run:

	./configure [<option> | -<option> | <option>=<value> ...]

The default values of feature options are taken from `default.cfg`.
Compiler flags and options dependent on compiler, OS and CPU type
are set in `ucw/autoconf.cfg`. Everything can be overriden by options
specified on the configure's command line, which have the highest
priority.

If you want to see the resulting set of options, take a look at
`obj/config.mk`.

Options specifying compiler/linker/debugger options can be also overriden
during compilation by `make <option>=<value>`. While it's also possible
to specify the other options in this way, it probably won't have the desired
effect, because configure also generates C include files containing the
options.

Installation options
--------------------
By default, the package is compiled to be installed and it chooses a
place where it will install. If you want to place it somewhere else,
set the `PREFIX` option:

	./configure PREFIX=/where/to/install

[[local_build]]
You can create a local compilation, which does not need to be
installed. All paths are set relative to a `run` directory and
programs expect to have current working directory set to it. However,
you can move the `run` directory around your filesystem as you like.
To compile it that way, turn on the `CONFIG_LOCAL` option (and do not
set `PREFIX`):

	./configure CONFIG_LOCAL

Examples
--------
  - `./configure -CONFIG_SHARED` will build the libraries statically.
  - `./configure -CONFIG_IMAGES` avoids building the image library.
  - `./configure CONFIG_UCW_ONLY -CONFIG_UCW_UTILS` compiles only the
    libucw library without its util programs.
  - `./configure CONFIG_XML PREFIX=$HOME/sherlock` includes a xml
    library and installs into your home directory.


lib/conf.c
/*
 *	UCW Library -- Reading of configuration files
 *
 *	(c) 2001 Robert Spalek <robert@ucw.cz>
 *	(c) 2003--2005 Martin Mares <mj@ucw.cz>
 *
 *	This software may be freely distributed and used according to the terms
 *	of the GNU Lesser General Public License.
 */

#include "lib/lib.h"
#include "lib/chartype.h"
#include "lib/fastbuf.h"
#include "lib/mempool.h"

#include "lib/conf.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <getopt.h>
#include <errno.h>

#define	BUFFER		1024
#define	MAX_LEVEL	8

static struct cfitem *cfsection;
struct mempool *cfpool;

#ifndef DEFAULT_CONFIG
#define DEFAULT_CONFIG NULL
#endif

byte *cfdeffile = DEFAULT_CONFIG;

static void CONSTRUCTOR
conf_init(void)
{
	cfpool = mp_new(4096);
}

void *
cfg_malloc(uns size)
{
	return mp_alloc(cfpool, size);
}

void *
cfg_malloc_zero(uns size)
{
	return mp_alloc_zero(cfpool, size);
}

byte *
cfg_strdup(byte *s)
{
	return mp_strdup(cfpool, s);
}

byte *
cfg_printf(char *fmt, ...)
{
	va_list args;
	va_start(args, fmt);
	byte *res = mp_vprintf(cfpool, fmt, args);
	va_end(args);
	return res;
}

void cf_register(struct cfitem *items)
{
	if(items[0].type!=CT_SECTION && items[0].type!=CT_INCOMPLETE_SECTION)
		die("cf_register: Invalid section type");
	items[0].var=cfsection;
	cfsection=items;
}

int cf_item_count(void)
{
	struct cfitem *sect, *item;
	int count = 0;
	for (sect = cfsection; sect; sect = sect->var)
		for (item = sect+1; item->type; item++)
			count++;
	return count;
}

struct cfitem *cf_get_item(byte *sect, byte *name)
{
	struct cfitem *item, *section;

	item=cfsection;
	while(item && strcasecmp(item->name,sect))
		item=item->var;
	if(!item)	/* unknown section */
		return NULL;
	section = item;

	for(item++; item->type && strcasecmp(item->name,name); item++);
	if (!item->type && section->type == CT_INCOMPLETE_SECTION)
		return NULL;

	return item;	/* item->type == 0 if not found */
}

struct unit {
	uns name;			/* One-letter name of the unit */
	uns num, den;			/* Fraction */
};

static const struct unit units[] = {
	{ 'd', 86400, 1 },
	{ 'h', 3600, 1 },
	{ 'k', 1000, 1 },
	{ 'm', 1000000, 1 },
	{ 'g', 1000000000, 1 },
	{ 'K', 1024, 1 },
	{ 'M', 1048576, 1 },
	{ 'G', 1073741824, 1 },
	{ '%', 1, 100 },
	{ 0, 0, 0 }
};

#if 0
static const struct unit *cf_lookup_unit(byte *value, byte *end, char **msg)
{
	if (end && *end) {
		if (end == value || end[1] || *end >= '0' && *end <= '9')
			*msg = "Invalid number";
		else {
			for (const struct unit *u=units; u->name; u++)
				if (u->name == *end)
					return u;
			*msg = "Invalid unit";
		}
	}
	return NULL;
}

static char cf_rngerr[] = "Number out of range";

byte *cf_parse_int(byte *value, uns *varp)
{
	char *msg = NULL;
	const struct unit *u;

	if (!*value)
		msg = "Missing number";
	else {
		errno = 0;
		char *end;
		uns x = strtoul(value, &end, 0);
		if (errno == ERANGE)
			msg = cf_rngerr;
		else if (u = cf_lookup_unit(value, end, &msg)) {
			u64 y = (u64)x * u->num;
			if (y % u->den)
				msg = "Number is not an integer";
			else {
				y /= u->den;
				if (y > 0xffffffff)
					msg = cf_rngerr;
				*varp = y;
			}
		} else
			*varp = x;
	}
	return msg;
}

byte *cf_parse_u64(byte *value, u64 *varp)
{
	char *msg = NULL;
	const struct unit *u;

	if (!*value)
		msg = "Missing number";
	else {
		errno = 0;
		char *end;
		u64 x = strtoull(value, &end, 0);
		if (errno == ERANGE)
			msg = cf_rngerr;
		else if (u = cf_lookup_unit(value, end, &msg)) {
			if (x > ~(u64)0 / u->num)
				msg = "Number out of range";
			else {
				x *= u->num;
				if (x % u->den)
					msg = "Number is not an integer";
				else
					*varp = x / u->den;
			}
		} else
			*varp = x;
	}
	return msg;
}

byte *cf_parse_double(byte *value, double *varp)
{
	char *msg = NULL;
	const struct unit *u;

	if (!*value)
		msg = "Missing number";
	else {
		errno = 0;
		char *end;
		double x = strtoul(value, &end, 0);
		if (errno == ERANGE)
			msg = cf_rngerr;
		else if (u = cf_lookup_unit(value, end, &msg))
			*varp = x * u->num / u->den;
		else
			*varp = x;
	}
	return msg;
}

byte *
cf_parse_ip(byte **p, u32 *varp)
{
  while (Cspace(**p))
    (*p)++;
  if (!**p)
    return "Missing IP address";
  uns x = 0;
  if (**p == '0' && *(*p + 1) | 32 == 'X')
    {
      errno = 0;
      x = strtoul(*p + 2, (char **)p, 16);
      if (errno == ERANGE || x > 0xffffffff)
        goto error;
    }
  else
    for (uns i = 0; i < 4; i++)
      {
        if (i)
          {
            while (Cspace(**p))
              (*p)++;
            if (*(*p)++ != '.')
              goto error;
          }
        while (Cspace(**p))
          (*p)++;
        errno = 0;
        uns y = strtoul(*p, (char **)p, 10);
        if (errno == ERANGE || y > 255)
          goto error;
        x = (x << 8) + y;
      }
  *varp = x;
  return NULL;
error:
  return "Invalid IP address";
}
#endif

byte *cf_set_item(byte *sect, byte *name, byte *value)
{
	struct cfitem *item;
	byte *msg=NULL;

	if (!*sect)
		return "Empty section name";
	item=cf_get_item(sect,name);
	if(!item)	/* ignore unknown section */
		return NULL;

	switch(item->type){
		case CT_INT:
			msg = cf_parse_int(value, (uns *) item->var);
			break;
		case CT_STRING:
			*((byte **) item->var) = cfg_strdup(value);
			break;
		case CT_FUNCTION:
			msg = ((ci_func) item->var)(item, cfg_strdup(value));
			break;
		case CT_DOUBLE:
			msg = cf_parse_double(value, (double *) item->var);
			break;
		case CT_U64:
			msg = cf_parse_u64(value, (u64 *) item->var);
			break;
		default:
			msg = "Unknown keyword";
	}

	return msg;
}

static int cf_subread(byte *filename,int level)
{
	int fd;
	struct fastbuf *b;
	byte def_section[BUFFER];
	int line;
	byte *msg=NULL;

	if(level>=MAX_LEVEL){
		log(L_ERROR,"Too many (%d) nested files when reading %s",level,filename);
		return 0;
	}
		
	fd=open(filename,O_RDONLY, 0666);
	if(fd<0){
		log(L_ERROR,"Cannot open configuration file %s: %m",filename);
		return 0;
	}
	b=bfdopen(fd,4096);

	def_section[0]=0;
	line=0;
	while(1){
		byte buf[BUFFER];
		byte *c;

		if(!bgets(b,buf,BUFFER))
			break;
		line++;

		c=buf+strlen(buf);
		while(c>buf && Cspace(c[-1]))
			*--c=0;
		c=buf;
		while(*c && Cspace(*c))
			c++;
		if(!*c || *c=='#')
			continue;

		if(*c=='['){
			strcpy(def_section,c+1);
			c=strchr(def_section,']');
			if(c){
				*c=0;
				if(c[1]){
					msg="Garbage after ]";
					break;
				}
			}else{
				msg="Missing ]";
				break;
			}

		}else{
			byte *sect,*name,*value;

			name=c;
			while(*c && !Cspace(*c))
				c++;
			while(*c && Cspace(*c))
				*c++=0;
			value=c;

			if(!strcasecmp(name,"include")){
				if(!cf_subread(value,level+1)){
					msg="Included from here";
					break;
				}
			}else{
				c=strchr(name,'.');
				if(!c)
					sect=def_section;
				else{
					sect=name;
					*c++=0;
					name=c;
				}

				msg=cf_set_item(sect,name,value);
			}
			if(msg)
				break;
		}

	}	/* for every line */

	if(msg)
		log(L_ERROR,"%s, line %d: %s",filename,line,msg);
	bclose(b);
	return !msg;
}

void cf_read(byte *filename)
{
	if(!cf_subread(filename,0))
		die("Reading config file %s failed",filename);
	cfdeffile = NULL;
}

static void cf_opt_S(void)
{
	byte *sect,*name,*value;
	byte *c;
	byte *msg=NULL;
	byte arg[strlen(optarg)+1];

	strcpy(arg, optarg);
	name = arg;
	c=strchr(name,'=');
	if(!c){
		msg="Missing argument";
		sect=value="";
	}else{
		*c++=0;
		value=c;

		c=strchr(name,'.');
		if(!c)
			sect="";
		else{
			sect=name;
			*c++=0;
			name=c;
		}

		if (cfdeffile)
			cf_read(cfdeffile);
		msg=cf_set_item(sect,name,value);
	}
	if(msg)
		die("Invalid command line argument -S%s.%s=%s: %s",sect,name,value,msg);

}

int cf_getopt(int argc,char * const argv[],
		const char *shortopts,const struct option *longopts,
		int *longindex)
{
	int res;
	static int other_options;

	do{
		res=getopt_long(argc,argv,shortopts,longopts,longindex);
		if(res == 'S' || res == 'C') {
			if(other_options)
				die("The -S and -C options must precede all other arguments");
		}
		if(res=='S'){
			cf_opt_S();
		}else if(res=='C'){
			cf_read(optarg);
		}else{
			/* unhandled option or end of options */
			if(cfdeffile)
				cf_read(cfdeffile);
			other_options++;
			return res;
		}
	}while(1);
}
